mesonbuild/build.py

```diff
@@ -2306,6 +2320,18 @@ def process_kwargs(self, kwargs):
             if rust_crate_type == 'proc-macro':
                 FeatureNew.single_use('Rust crate type "proc-macro"', '0.62.0', self.subproject)
 
+        if 'rust_dependency_map' in kwargs:
+            rust_dependency_map = kwargs['rust_dependency_map']
+            if isinstance(rust_dependency_map, dict):
+                for (k, v) in rust_dependency_map.items():
+                    if not isinstance(k, str) or not isinstance(v, str):
+                        raise InvalidArguments(f'Invalid rust_dependency_map "{rust_dependency_map}": must be a string to string dictionary.')
+
+                self.rust_dependency_map = rust_dependency_map
+            else:
+                raise InvalidArguments(f'Invalid rust_dependency_map "{rust_dependency_map}": must be a dictionary.')
```

Comment:
That can be moved into base class instead of duplicating. Also if you do `self.rust_dependency_map = kwargs.get('rust_dependency_map', {})` then you can remove the hasattr() in post_init().

Ideally the typechecking should be done in interpreter, but I think @dcbaker's PR for that has not landed yet.

Nitpicking: no need to check key type, meson language only allow string keys in dict already. `if any(not isinstance(v, str) for v in rust_dependency_map.values())`.

Reply:
> That can be moved into base class instead of duplicating. Also if you do self.rust_dependency_map = kwargs.get('rust_dependency_map', {}) then you can remove the hasattr() in post_init().

Which base class? This is currently mirroring how `rust_crate_type` is handled but I can do it differently too :)

> Ideally the typechecking should be done in interpreter, but I think @dcbaker's PR for that has not landed yet.

Yeah I don't think any of the library types are covered by that right now, that's what I was checking first but couldn't find anything.

> Nitpicking: no need to check key type, meson language only allow string keys in dict already. if any(not isinstance(v, str) for v in rust_dependency_map.values()).

Nicer, thanks :)

Reply:
> Which base class? This is currently mirroring how `rust_crate_type` is handled but I can do it differently too :)

BuildTarget, it's true that not all subclasses of BuildTarget needs those rust attributes, but does not really hurt neither. Maybe we need an intermediary LibraryTarget class, but IMHO that's overkill for now.

Reply:
`BuildTarget` makes more sense even. This shouldn't only be done for library targets but also for executables. Anything that can have a Rust dependency.